### PR TITLE
fix: store strong references of task objects 

### LINKED ISSUE #63: Async reduce UDF returns Runtime error for asynchronous generator
**Describe the bug**
While using the async reduce UDF if running a function with a high I/O delay or at high input message throughput, the pod seems to crash with the following error. In my case the UDF makes an HTTP request to a server and sends the response to the sink.
```
future: <Task finished name='Task-25619' coro=<<async_generator_athrow without __name__>()>exception=RuntimeError('aclose(): asynchronous generator is already running')>
RuntimeError: aclose(): asynchronous generator is already running.
```
This is reproducible, and depending on the input RPU and the response time, this might occur from the start of processing or after sometime. 

**To Reproduce**

Steps to reproduce using the attached files 
1. Change the attached UDF code at line 28 to point to any web server and change the response in line 15 accordingly
2. Build the image and create a new pipeline using the YAML file.
3. The issue is observed at a high value of RPU (>100) or high response delay from the server
4. Monitor the UDF vertex logs for the error message

**Expected behavior**
Messages should be processed or graceful termination with proper exception handling

**Screenshots**

Screenshot 2023-03-06 at 9 40 40 AM
Screenshot 2023-03-06 at 9 42 10 AM

**Environment (please complete the following information):**
- pynumaflow = "~0.3.2"

**Additional context**
Reference files 
[aclose.zip](https://github.com/numaproj/numaflow-python/files/10901046/aclose.zip)

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by kohlisid**
Fixes #63 
Save strong references of tasks/futures so that they are not lost mid execution due to GC.

### CODE REVIEW

File: pynumaflow/function/server.py [hunk 1]
```diff
@@ -242,6 +245,8 @@ async def __async_reduce_handler(self, interval_window, request_iterator: AsyncI
                 task = asyncio.create_task(
                     self.__invoke_reduce(key, riter, Metadata(interval_window=interval_window))
                 )
+                self.background_tasks.add(task)
```
Comment (ab93) [inline pynumaflow/function/server.py:248]: Add a quick comment here maybe mentioning why we are doing this?

**Review by ab93 — APPROVED** [review]
Small comment

File: pynumaflow/function/server.py [hunk 2]
```diff
@@ -109,6 +109,7 @@ def __init__(
         self._max_message_size = max_message_size
         self._max_threads = max_threads
         self.cleanup_coroutines = []
+        self.background_tasks = set()
```
Comment (vigith) [inline pynumaflow/function/server.py:112]: could you please add a comment around this? please reference the weakref in asyncio

File: pynumaflow/function/server.py (hunk 1 above)
Comment (vigith) [inline pynumaflow/function/server.py:248]: try without doing this and see whether we endup in weakref.. if yes, we will have to call this out in our doc.

Comment (vigith) [reply pynumaflow/function/server.py:248]: the second save is concerning as users unfamiliar with asyncio are bound to miss.

Comment (kohlisid) [reply pynumaflow/function/server.py:248]: The observation from the testing indicates that just using the second save works for us, but using only the first save continues to throw the error. 
Digging in to confirm the expectations for weak references in python.

**Comment by KeranYang** [comment]
@kohlisid FYI. The codecov failure could be resolved by https://github.com/numaproj/numaflow-python/pull/67#discussion_r1138831711

File: pynumaflow/function/server.py [hunk 3]
```diff
@@ -220,6 +221,11 @@ async def ReduceFn(
             self.__async_reduce_handler(interval_window, request_iterator)
         )
 
+        # Save a reference to the result of this function, to avoid a
+        # task disappearing mid-execution.
+        self.background_tasks.add(response_task)
```
Comment (vigith) [inline pynumaflow/function/server.py:226]: why do we still need this?

Comment (vigith) [reply pynumaflow/function/server.py:226]: won't the second reference saving solve this?

File: pynumaflow/function/server.py (hunk 1 above)
Comment (kohlisid) [reply pynumaflow/function/server.py:256]: The references are not hierarchical in nature for the coroutines, child is not automatically referred by the parent.

**Pull request merged**